### PR TITLE
Shrestha/backend experiments

### PR DESCRIPTION
**Pull request by shresthamalik**
Implemented an initial skeleton for Backend Handling

1. Added a backend manager
            - Maintains a cache/map of backends and their synchronization primitives
            - Support for set backend, supported backends etc

2. nGraphEncapsulate Op
            - Gets a raw pointer for the backend from backend manager, in the compute function to run the computation
            - The op does not hold onto the backend pointer, ie. no data member for the backend   
            - The op has a member backend_name which is assigned a value in the Constructor() for the op, from the "_ngraph_backend" attribute that is added to it while creating the Node in Encapsulate Clusters

3. Rewrite Passes
           - In mark_for_clustering we add "_ngraph_backend" attribute to each node that is marked for clustering
           - In Assign Clusters we do not contract an edge for which src and dest are assigned different backend are clustered together
            - In Encapsulate Clusters we do a check (along with the device check) reaffirming the above. While creating the nGraphEncapsulateOp, it is assigned "_ngraph_backend"

### CODE REVIEW

File: src/ngraph_backend_manager.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+struct Backend {
+  unique_ptr<ng::runtime::Backend> backend_ptr;
+  mutex backend_mutex;
+};
+
+class BackendManager {
+ public:
+  // Returns the backend name currently set
+  static string GetCurrentlySetBackendName() {
+    NGRAPH_VLOG(5) << "Getting backend";
```
Comment (sayantan-nervana) [inline src/ngraph_backend_manager.h]: Can we say: `Reading backend name`

File: src/ngraph_backend_manager.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+using namespace std;
+namespace ng = ngraph;
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+struct Backend {
+  unique_ptr<ng::runtime::Backend> backend_ptr;
+  mutex backend_mutex;
```
Comment (sayantan-nervana) [inline src/ngraph_backend_manager.h]:
```
mutex backend_mutex;
unique_ptr<ng::runtime::Backend> backend_ptr GUARDED_BY (backend_mutex);
```

Not sure if gcc honours `GUARDED_BY` annotation though

File: src/ngraph_backend_manager.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  static ng::runtime::Backend* GetBackend(const string& backend_name);
+
+  // LockBackend
+  static void LockBackend(const string& backend_name);
+
+  // UnlockBackend
+  static void UnlockBackend(const string& backend_name);
+
+ private:
+  static string ng_backend_name_;  // currently set backend name
```
Comment (sayantan-nervana) [inline src/ngraph_backend_manager.h]: `GUARDED_BY(ng_backend_name_mutex_)`

File: src/ngraph_encapsulate_op.cc [hunk 4]
```diff
@@ -478,13 +470,14 @@ class NGraphEncapsulateOp : public OpKernel {
             ctx, false,
             errors::Internal("Error in executing the nGraph computation\n"));
       }
+      BackendManager::UnlockBackend(m_op_backend_name);
```
Comment (sayantan-nervana) [inline src/ngraph_encapsulate_op.cc]: Slightly unsafe that we are locking/unlocking by hand. But there is a catch-all exception here, so no error should happen which would leave the mutex in a locked state.

File: test/graph_rewrites/backend_manager_test.cc [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+namespace testing {
+
+#define ASSERT_OK(x) ASSERT_EQ((x), ::tensorflow::Status::OK());
+#define ASSERT_NOT_OK(x) ASSERT_NE((x), ::tensorflow::Status::OK());
+
+/*
+These tests test the Backend Handling by the bridge
+Since the backend is set globaly the setting of NGRAPH_TF_BACKEND might cause
+some tests to fail
+For e.g. "BackendCLustering" test would fail If the NGRAPH_TF_BACKEND is set to
```
Comment (sayantan-nervana) [inline test/graph_rewrites/backend_manager_test.cc]: `CLustering`->`Clustering`

File: src/ngraph_backend_manager.h (hunk 3 above)
Comment (shresthamalik) [reply src/ngraph_backend_manager.h]: Yes, was trying to add just that getting compilation errors

File: src/ngraph_assign_clusters.cc [hunk 6]
```diff
@@ -425,10 +460,27 @@ Status AssignClusters(Graph* graph) {
...
+                       << dst->name() << "[" << edge->dst_input() << "]@"
+                       << dst_index;
         continue;
       }
 #endif
 
+      // check if the edge can be constracted with respect to backend
+      bool is_backend_ok = false;
+      TF_RETURN_IF_ERROR(
+          CanContractEdgeBackendCheck(edge, cluster_map, is_backend_ok));
```
[the remark this answers is not in the document]
Comment (shresthamalik) [reply src/ngraph_assign_clusters.cc]: Yes, was just maintaining the same api as CanContractDeadness and future scoping it ... Can move to boolean

File: src/ngraph_backend_manager.h (hunk 1 above)
Comment (shresthamalik) [reply src/ngraph_backend_manager.h]: This was just a debug log. Removing this :)

File: src/ngraph_encapsulate_op.cc (hunk 4 above)
Comment (shresthamalik) [reply src/ngraph_encapsulate_op.cc]: Yes, and the next level of refactoring/design will get rid of this, and move most of this to Backend class

**Comment by shresthamalik** [comment]
All the 21 models pass with this 85268dccc90d01b58e464ea4031445f3f36b31cb revision of this PR. Ran with ng-tf master , ngraph version aac6ee2a2ba1d1a1aa82a9635389c593943ae91d

File: src/ngraph_assign_clusters.cc [hunk 7]
```diff
@@ -91,12 +91,42 @@ namespace {
...
   int index;
   std::set<tensorflow::Node*> nodes;
+  std::string backend;
 #if !defined(NGRAPH_TF_DISABLE_DEADNESS_CHECK)
   std::string predicate_string;
   std::set<const Edge*> outgoing_edges;
 #endif
 };
 
+Status InitialiseNodeBackend(Node* node, string* backend) {
```
Comment (kanvi-nervana) [inline src/ngraph_assign_clusters.cc]: Minor: Initialise to Initialize :)

File: src/ngraph_assign_clusters.cc [hunk 8]
```diff
@@ -91,12 +91,42 @@ namespace {
...
+  NGRAPH_VLOG(5) << "Initialize Node Backend " << node->name();
+  if (!HasNodeAttr(node->def(), "_ngraph_backend")) {
+    *backend = "HOST";
+    return Status::OK();
+  }
+  NGRAPH_VLOG(5) << "Should have been assigned Node Backend " << node->name();
+  TF_RETURN_IF_ERROR(GetNodeBackend(node, backend));
+  // TF_RETURN_IF_ERROR(GetNodeAttr(node->attrs(), "_ngraph_backend", backend));
```
Comment (kanvi-nervana) [inline src/ngraph_assign_clusters.cc]: Can you remove the commented out code?

File: src/ngraph_assign_clusters.cc [hunk 9]
```diff
@@ -303,9 +333,14 @@ Status AssignClusters(Graph* graph) {
     int new_index = gc.NewNode();
     cluster_map[node] = std::make_shared<Cluster>();
     cluster_map[node]->index = new_index;
+    string backend;
+    TF_RETURN_IF_ERROR(InitialiseNodeBackend(node, &backend));
```
Comment (kanvi-nervana) [inline src/ngraph_assign_clusters.cc]: Initialise to Initialize

File: src/ngraph_assign_clusters.cc [hunk 10]
```diff
@@ -425,10 +460,27 @@ Status AssignClusters(Graph* graph) {
           CanContractEdgeDeadnessCheck(edge, cluster_map, is_deadness_ok));
       if (!is_deadness_ok) {
         // do not contract, src and dst node cannot be in the same cluster
+        NGRAPH_VLOG(5) << "Skipping (deadness not ok): " << src->name() << "["
```
Comment (kanvi-nervana) [inline src/ngraph_assign_clusters.cc]: Thanks for adding these logs.

File: src/ngraph_assign_clusters.cc [hunk 11]
```diff
@@ -425,10 +460,27 @@ Status AssignClusters(Graph* graph) {
...
+                       << edge->src_output() << "]@" << src_index << " -> "
+                       << dst->name() << "[" << edge->dst_input() << "]@"
+                       << dst_index;
         continue;
       }
 #endif
 
+      // check if the edge can be constracted with respect to backend
```
Comment (kanvi-nervana) [inline src/ngraph_assign_clusters.cc]: constracted to contracted

File: src/ngraph_backend_manager.h (hunk 2 above)
Comment (kanvi-nervana) [reply src/ngraph_backend_manager.h]: I think it does.

File: src/ngraph_encapsulate_clusters.cc [hunk 12]
```diff
@@ -38,6 +38,8 @@
 #include "ngraph_cluster_manager.h"
 #include "ngraph_encapsulate_clusters.h"
 #include "ngraph_log.h"
+#include "ngraph_mark_for_clustering.h"
```
Comment (kanvi-nervana) [inline src/ngraph_encapsulate_clusters.cc]: #include "ngraph_mark_for_clustering.h" is added twice

File: src/ngraph_encapsulate_op.cc [hunk 13]
```diff
@@ -515,29 +508,29 @@ class NGraphEncapsulateOp : public OpKernel {
...
   NgFunctionIOCache m_ng_function_output_cache_map;
   NGraphFreshnessTracker* m_freshness_tracker;
   int m_ngraph_cluster;
   std::vector<bool> m_input_is_static;
-
-  static std::string s_ng_backend_name;
-  static mutex s_ng_backend_mutex;
   std::mutex m_compute_lock;
+  string m_op_backend_name;
+  // static std::weak_ptr<ng::runtime::Backend> s_ng_backend_wptr;
```
Comment (kanvi-nervana) [inline src/ngraph_encapsulate_op.cc]: Why do we need the commented out code?

File: test/CMakeLists.txt [hunk 14]
```diff
@@ -56,6 +56,7 @@ set(SRC
     conversions.cpp
     graph_rewrites/assign_clusters.cc
     graph_rewrites/deadness_test.cc
+    graph_rewrites/backend_manager_test.cc
```
Comment (sindhu-nervana) [inline test/CMakeLists.txt]: Minor: maybe move line `graph_rewrites/backend_manager_test.cc` before  `graph_rewrites/deadness_test.cc` just to keep it alphabetical like others?

**Pull request merged**